scrape/scrape.go

```diff
@@ -932,77 +932,99 @@ mainLoop:
-				time.Since(last).Seconds(),
-			)
+		select {
+		case <-sl.parentCtx.Done():
+			close(sl.stopped)
+			return
+		case <-sl.ctx.Done():
+			break mainLoop
+		case <-ticker.C:
 		}
+	}
 
-		b := sl.buffers.Get(sl.lastScrapeSize).([]byte)
-		buf := bytes.NewBuffer(b)
+	close(sl.stopped)
 
-		contentType, scrapeErr := sl.scraper.scrape(scrapeCtx, buf)
-		cancel()
+	if !sl.disabledEndOfRunStalenessMarkers {
+		sl.endOfRunStaleness(last, ticker, interval)
+	}
+}
 
-		if scrapeErr == nil {
-			b = buf.Bytes()
-			// NOTE: There were issues with misbehaving clients in the past
-			// that occasionally returned empty results. We don't want those
-			// to falsely reset our buffer size.
-			if len(b) > 0 {
-				sl.lastScrapeSize = len(b)
-			}
-		} else {
-			level.Debug(sl.l).Log("msg", "Scrape failed", "err", scrapeErr.Error())
-			if errc != nil {
-				errc <- scrapeErr
-			}
-		}
+// scrapeAndReport performs a scrape and then appends the result to the storage
+// together with reporting metrics, all using the same appender.
+func (sl *scrapeLoop) scrapeAndReport(interval, timeout time.Duration, last time.Time, errc chan<- error) time.Time {
+	var (
+		start             = time.Now()
+		scrapeCtx, cancel = context.WithTimeout(sl.ctx, timeout)
+	)
 
-		// A failed scrape is the same as an empty scrape,
-		// we still call sl.append to trigger stale markers.
-		total, added, seriesAdded, appErr := sl.append(b, contentType, start)
-		if appErr != nil {
-			level.Debug(sl.l).Log("msg", "Append failed", "err", appErr)
-			// The append failed, probably due to a parse error or sample limit.
-			// Call sl.append again with an empty scrape to trigger stale markers.
-			if _, _, _, err := sl.append([]byte{}, "", start); err != nil {
-				level.Warn(sl.l).Log("msg", "Append failed", "err", err)
-			}
-		}
+	// Only record after the first scrape.
+	if !last.IsZero() {
+		targetIntervalLength.WithLabelValues(interval.String()).Observe(
+			time.Since(last).Seconds(),
+		)
+	}
 
-		sl.buffers.Put(b)
+	b := sl.buffers.Get(sl.lastScrapeSize).([]byte)
+	buf := bytes.NewBuffer(b)
 
-		if scrapeErr == nil {
-			scrapeErr = appErr
-		}
+	contentType, scrapeErr := sl.scraper.scrape(scrapeCtx, buf)
+	cancel()
 
-		if err := sl.report(start, time.Since(start), total, added, seriesAdded, scrapeErr); err != nil {
-			level.Warn(sl.l).Log("msg", "Appending scrape report failed", "err", err)
+	if scrapeErr == nil {
+		b = buf.Bytes()
+		// NOTE: There were issues with misbehaving clients in the past
+		// that occasionally returned empty results. We don't want those
+		// to falsely reset our buffer size.
+		if len(b) > 0 {
+			sl.lastScrapeSize = len(b)
+		}
+	} else {
+		level.Debug(sl.l).Log("msg", "Scrape failed", "err", scrapeErr.Error())
+		if errc != nil {
+			errc <- scrapeErr
 		}
-		last = start
+	}
 
-		select {
-		case <-sl.parentCtx.Done():
-			close(sl.stopped)
+	app := sl.appender()
+	var err error
+	defer func() {
```

Comment:
There's no early returns, so I don't see a need to use a defer here compared to putting this code at the end of the function - which would be clearer.

Reply:
We guard ourselves against potential panic()